### ISSUE TITLE
Nanoset reuses same shuffling for repeated data (epochs)

### ISSUE THREAD

**Issue by Lauler** [huggingface/nanotron#237]
Nanoset's index builder does not re-shuffle dataset and sample indices within epochs when training secondary, third, etc epochs. It instead concatenates a copy of the same indices for any repeated data. See:

https://github.com/huggingface/nanotron/blob/51ca40bc5e1b1f5dcb55eaeb0b6f86dda03f3979/src/nanotron/data/nanoset.py#L114-L124

I would suggest changing this so the shuffling is applied within each epoch before concatenating the data. Here's the edited version I use:

```python
# Shuffle the indices within each epoch and concatenate them
r = np.random.RandomState(self.random_seed)
epoch_random_seeds = r.randint(0, 2**32 - 1, num_epochs)
dataset_indices = []
dataset_sample_indices = []
for i in range(num_epochs):
    # Shuffle the sample and dataset indices in epoch with same seed
    numpy_random_state = np.random.RandomState(epoch_random_seeds[i])
    numpy_random_state.shuffle(dataset_index)
    numpy_random_state = np.random.RandomState(epoch_random_seeds[i])
    numpy_random_state.shuffle(dataset_sample_index)

    dataset_indices.append(dataset_index)
    dataset_sample_indices.append(dataset_sample_index)

# Concatenate the within-epoch shuffled indexes
dataset_index = np.concatenate(dataset_indices)
dataset_sample_index = np.concatenate(dataset_sample_indices)
```

If you think this is reasonable I can submit a pull request.

**Comment by TJ-Solergibert** [comment]
Yes! I [warned](https://github.com/huggingface/nanotron/blob/51ca40bc5e1b1f5dcb55eaeb0b6f86dda03f3979/docs/nanoset.md?plain=1#L110) this in the little docs I wrote a while back. 

Looks good to me, make sure that the unit tests are passing and try to submit a PR!
